network/p2p/connManager.go

```diff
@@ -87,3 +98,47 @@ func (c ConnManager) logConnectionUpdate(n network.Network, con network.Conn, lo
 		Int("total_connections", len(n.Conns())).
 		Msg(logMsg)
 }
+
+// ProtectPeer increments the stream setup count for the peer.ID
+func (c *ConnManager) ProtectPeer(id peer.ID) {
+	c.streamSetupMapLk.Lock()
+	defer c.streamSetupMapLk.Unlock()
```

Comment:
This can definitely wait for another PR but: for better maintainability would be great if we either encapsulate this entire logic into a separate struct (e.g., `StreamSetupProtector`, or use [atomic increment](https://gobyexample.com/atomic-counters). The reason I suggest this is to deter having a single-purpose mutex lingering around a more general-purpose struct. In other words, `streamSetupMapLK` only protects this map, however, there is no guarantee that any other component of this `struct` does not utilize it for another purpose.